nautobot/utilities/forms/fields.py

```diff
@@ -666,3 +666,41 @@ def to_python(self, value):
         except ValueError as error:
             raise ValidationError(error)
         return super().to_python(value)
+
+
+class NaturalKeyMultipleChoiceField(django_filters.fields.ModelMultipleChoiceField):
+    """
+    Field to support matching an object's `pk` or `slug` field depending on the value
+    supplied. Faises ValidationError if neither field matches. Matches `pk` on the
+    `iexact` lookup expression or `slug` on `icontains`.
+    """
+
+    def _check_values(self, values):
+        """
+        This method overloads the grandparent method in `django.forms.models.ModelMultipleChoiceField`,
+        re-using some of that method's existing logic and adding support for coupling this field with
+        multiple model fields.
+        """
+        # deduplicate given values to avoid creating many querysets or
+        # requiring the database backend deduplicate efficiently.
+        try:
+            values = frozenset(values)
+        except TypeError:
+            # list of lists isn't hashable, for example
+            raise ValidationError(
+                self.error_messages["invalid_list"],
+                code="invalid_list",
+            )
+        for item in values:
+            qs = self.queryset.filter(Q(pk__iexact=str(item)) | Q(slug__icontains=str(item)))
+            if not qs.exists():
+                raise ValidationError(
+                    self.error_messages["invalid_choice"],
+                    code="invalid_choice",
+                    params={"value": item},
+                )
+        query = Q()
+        for item in values:
+            query |= Q(pk__iexact=str(item)) | Q(slug__icontains=str(item))
```

Comment:
Naively I would have expected something more like:

```python
query = Q(pk__in(values)) | Q(slug__in(values))
```

Does that approach not work? Are there any concerns as to the efficiency of this query with the current approach?

Reply:
It looks like django does a pretty good job optimizing both queries but with a larger dataset the `__in` query is a lot faster. I'll need to make sure I only pass valid UUIDs to the `pk__in` query so that will require a little more memory/processing on the python side. Using the `pk__iexact` query converts the left and right side values of the evaluation to strings so I didn't have a problem comparing UUID to any random string.

Using `__in`:
```sql
SELECT "fields" FROM "table" WHERE ("table"."id" IN (value1, value2, value3) OR "table"."slug" IN (value1, value2, value3)) ORDER BY "table"."_name" ASC
```


`__in` query analysis:
```
Sort  (cost=12.68..12.73 rows=20 width=3038) (actual time=0.052..0.053 rows=41 loops=1)
Planning Time: 0.118 ms\nExecution Time: 0.065 ms
```

Using ORed Q queries:
```sql
SELECT "fields" FROM "table" WHERE ("table"."id" = value1 OR "table"."slug" = value1 OR "table"."id" = value2 OR "table"."slug" = value2 OR "table"."id" = value3 OR "table"."slug" = value3) ORDER BY "table"."_name" ASC
```

ORed Q query analysis:
```
Sort  (cost=141.83..141.88 rows=20 width=3038) (actual time=0.099..0.100 rows=41 loops=1)
Planning Time: 5.481 ms\nExecution Time: 0.535 ms
```